web-ui/src/app/services/treebank.service.ts

```diff
@@ -47,196 +50,346 @@ interface ConfiguredTreebanksResponse {
             minValue?: number | Date,
             maxValue?: number | Date,
         }[],
-        'multioption': boolean
+        'multioption'?: boolean
     };
-}
+};
+
+type UploadedTreebankResponse = {
+    email: string,
+    id: string
+    processed: string,
+    public: '1'|'0',
+    title: string,
+    uploaded: string,
+    user_id: string
+};
+
+type UploadedTreebankMetadataResponse = {
+    id: string,
+    treebank_id: string,
+    field: string,
+    type: 'text' | 'int' | 'date',
+    facet: 'checkbox' | 'slider' | 'date_range',
+    min_value: string | null,
+    max_value: string | null,
+    show: '1' | '0'
+};
+
+// not quite sure what this is yet
+type UploadedTreebankShowResponse = {
+    basex_db: string,
+    nr_sentences: string,
+    nr_words: string,
+    slug: string,
+    title: string
+};
+
 @Injectable()
 export class TreebankService {
+    private loading = true;
+    private data: ConfiguredTreebanks;
+    public treebanks = new ReplaySubject<{state: ConfiguredTreebanks, origin: 'init'|'url'|'user'}>(1);
 
     constructor(private configurationService: ConfigurationService, private http: HttpClient) {
-    }
-
-    private async getConfiguredTreebanks() {
-        if (configuredTreebanks) {
-            return configuredTreebanks;
-        }
+        (async () => {
```

Comment:
Hier liep ik tegenaan toen ik de unit tests weer aan de praat probeerde te krijgen, dit is erg ingewikkeld om te testen. Hier lopen nu ook twee dingen door elkaar heen: wat de gebruiker wel/niet heeft geselecteerd, en de logica om treebanks op te halen. Het is handiger als die twee zaken ook in twee verschillende services zijn gespecificeerd. Zodat deze service zich weer enkel bezig kan houden met het ophalen van treebanks en die andere service dan zich alleen hoeft te bekommeren om wat er wel of niet geselecteerd is. Dat is veel makkelijker te testen en denk ik ook een stuk begrijpelijker.

Reply:
Hier nog even over, hoe zou dit (ongeveer) beter op te zetten zijn van uit een test oogpunt?
Het lazy ophalen van treebank info zorgt er immers ook voor dat de rest van applicatie ineens overal rekening moet houden met banks die misschien nog undefined zijn. 
Is het splitsen naar wat meer granulaire getters afdoende?

Ik zal de logica wel meer opsplitsen en waar mogelijk niet langer allemaal async maken.